ruby/ql/lib/codeql/ruby/security/UrlConcatenation.qll

```diff
@@ -0,0 +1,84 @@
+/**
+ * Provides a class for detecting string concatenations involving
+ * the characters `?` and `#`, which are considered sanitizers for
+ * the URL redirection queries.
+ */
+
+import ruby
+import codeql.ruby.StringOps
+import codeql.ruby.StringConcatenation
+
+/**
+ * Holds if the string value of `nd` prevents anything appended after it
+ * from affecting the hostname or path of a URL.
+ *
+ * Specifically, this holds if the string contains `?` or `#`.
+ */
+private predicate hasSanitizingSubstring(DataFlow::Node nd) {
+  nd.getConstantValue().getStringlikeValue().regexpMatch(".*[?#].*")
+  or
+  hasSanitizingSubstring(StringConcatenation::getAnOperand(nd))
+  or
+  hasSanitizingSubstring(nd.getAPredecessor())
+}
+
+/**
+ * Holds if data that flows from `source` to `sink` cannot affect the
+ * path or earlier part of the resulting string when interpreted as a URL.
+ *
+ * This is considered as a sanitizing edge for the URL redirection queries.
+ */
+predicate sanitizingPrefixEdge(DataFlow::Node source, DataFlow::Node sink) {
```

Comment:
JS and Go use this for their request forgery queries - so a URL built from user-input is considered safe only if the protocol, hostname, and path are all safe rather than just the protocol and hostname. In our documentation for SSRF we only mention cases where the host/protocol is targeted - but maybe this is too permissive?

For caution's sake, I think it's probably better if we change SSRF to use `sanitizingPrefixEdge` rather than `hostnameSanitizingPrefixEdge`.